### PR TITLE
Integration tests capable of testing multiple running reaper instances

### PR DESCRIPTION
**Pull request by michaelsembwever**
Work in progress still.

Make integration tests capable of testing multiple running reaper instances

### CODE REVIEW

File: src/main/java/com/spotify/reaper/service/SchedulingManager.java [hunk 1]
```diff
@@ -103,79 +104,131 @@ public void run() {
...
-      if (schedule.getState() == RepairSchedule.State.PAUSED) {
-        LOG.info("Repair schedule '{}' is paused", schedule.getId());
-        startNewRun = false;
-      } else {
-        Optional<RepairUnit> fetchedUnit =
-            context.storage.getRepairUnit(schedule.getRepairUnitId());
-        if (!fetchedUnit.isPresent()) {
-          LOG.warn("RepairUnit with id {} not found", schedule.getRepairUnitId());
-          return false;
+  private boolean manageSchedule(RepairSchedule schedule_) {
```
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/SchedulingManager.java]: MINOR Rename this local variable to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00117)

**Comment by tlpsonarqube** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [CassandraStorage.java#L491](https://github.com/thelastpickle/cassandra-reaper/blob/522f1aa5e6b03ae964e1ffeaa83b66384a1bb50f/src/main/java/com/spotify/reaper/storage/CassandraStorage.java#L491): Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)
1. ![MAJOR][MAJOR] [CassandraStorage.java#L796](https://github.com/thelastpickle/cassandra-reaper/blob/522f1aa5e6b03ae964e1ffeaa83b66384a1bb50f/src/main/java/com/spotify/reaper/storage/CassandraStorage.java#L796): Remove this "return" statement or make it conditional. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1751)
1. ![MINOR][MINOR] [SchedulingManager.java#L8](https://github.com/thelastpickle/cassandra-reaper/blob/522f1aa5e6b03ae964e1ffeaa83b66384a1bb50f/src/main/java/com/spotify/reaper/service/SchedulingManager.java#L8): Remove this unused import 'com.spotify.reaper.core.Cluster'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/spotify/reaper/service/SchedulingManager.java [hunk 2]
```diff
@@ -103,79 +104,131 @@ public void run() {
...
+                  List<UUID> currentRunHistory = context.storage
+                          .getRepairSchedule(schedule.getId()).get().getRunHistory();
+
+                  if (equalRunHistories(schedule, currentRunHistory)) {
+                    boolean result = context.storage.updateRepairSchedule(
+                          schedule.with()
+                            .runHistory(newRunHistory)
+                            .build(schedule.getId()));
+
+                    if (result && !repairRunAlreadyScheduled(schedule, repairUnit)) {
```
Comment (michaelsembwever) [inline src/main/java/com/spotify/reaper/service/SchedulingManager.java]: this `repairRunAlreadyScheduled(..)` needs to be removed.

**Comment by michaelsembwever** [comment]
@adejanovski 
 this is ready for further testing.

currently the repair and incremental repair scenarios break (most of the time) when running ReaperCassandraIT due to valid concurrency problems¹. 
(when they fail note that they can also cause other scenarios to also fail.)

the problem exceerbates as you add ccm nodes, increase RF, and increase concurrency in ReaperCassandraIT (see line 50).

the concurrency problems do appear to be those i raised concerns about earlier in https://github.com/thelastpickle/cassandra-reaper/pull/49#discussion_r115172770 and https://github.com/thelastpickle/cassandra-reaper/pull/49#discussion_r115169095

File: src/test/java/com/spotify/reaper/acceptance/BasicSteps.java [hunk 3]
```diff
@@ -767,7 +767,7 @@ public void we_wait_for_at_least_segments_to_be_repaired(int nbSegmentsToBeRepai
           assertEquals(Response.Status.OK.getStatusCode(), response.getStatus());
           String responseData = response.readEntity(String.class);
           RepairRunStatus run = SimpleReaperClient.parseRepairRunStatusJSON(responseData);
-          return nbSegmentsToBeRepaired == run.getSegmentsRepaired();
+          return nbSegmentsToBeRepaired <= run.getSegmentsRepaired();
```
Comment (michaelsembwever) [inline src/test/java/com/spotify/reaper/acceptance/BasicSteps.java]: :+1:

**Comment by michaelsembwever** [comment]
rebased off `ft-reaper-improvements-final`

**Comment by michaelsembwever** [comment]
The work in this branch has become significantly large.
The testing of multiple reaper instances running on top of a c* storage smoked out a lot of errors in the code, design, and rest interface.

The travis build now separates jobs, to see easily where distributed tests are failing.
Separate test hardware will need to be used for more extensive integration testing. Travis can only handle tests running a few reaper instances.

The distributed tests still fail sometimes as the design of the rest interface is inherently not consistent, and the use of strict consistency and leader election within the code does not address this. 
But it is close now, and some further bug fixing could get it to consistent 100% pass rate.

File: .travis.yml [hunk 4]
```diff
@@ -1,15 +1,23 @@
...
     - LOCAL_JMX=no
+    - TEST=""
+    - GRIM_MIN=1
+    - GRIM_MAX=1
   matrix:
-    - CASSANDRA_VERSION=2.1.16
-    - CASSANDRA_VERSION=2.2.8
-    - CASSANDRA_VERSION=3.0.9
-    - CASSANDRA_VERSION=3.10
+#    - CASSANDRA_VERSION=2.1.18
```
Comment (michaelsembwever) [inline .travis.yml]: @adejanovski i could not get 2.1.x to work, not even without multiple reaper instances…
the JmxProxy could not connect to *any* of ccm nodes.
i have no idea why yet, but suspect it is just a my machine problem.
would be great if you could confirm this!

(and without 2.1.x there's no sonar report)

File: src/main/java/com/spotify/reaper/storage/CassandraStorage.java [hunk 5]
```diff
@@ -127,11 +135,11 @@ private void prepareStatements(){
...
     getRepairUnitPrepStmt = session.prepare("SELECT * FROM repair_unit_v1 WHERE id = ?");
-    insertRepairSegmentPrepStmt = session.prepare("INSERT INTO repair_run(id, segment_id, repair_unit_id, start_token, end_token, segment_state, coordinator_host, segment_start_time, segment_end_time, fail_count) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)");
-    getRepairSegmentPrepStmt = session.prepare("SELECT id,repair_unit_id,segment_id,start_token,end_token,segment_state,coordinator_host,segment_start_time,segment_end_time,fail_count FROM repair_run WHERE id = ? and segment_id = ?");
+    insertRepairSegmentPrepStmt = session.prepare("INSERT INTO repair_run(id, segment_id, repair_unit_id, start_token, end_token, segment_state, coordinator_host, segment_start_time, segment_end_time, fail_count) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)").setConsistencyLevel(ConsistencyLevel.QUORUM);
+    getRepairSegmentPrepStmt = session.prepare("SELECT id,repair_unit_id,segment_id,start_token,end_token,segment_state,coordinator_host,segment_start_time,segment_end_time,fail_count FROM repair_run WHERE id = ? and segment_id = ?").setConsistencyLevel(ConsistencyLevel.QUORUM);
```
Comment (michaelsembwever) [inline src/main/java/com/spotify/reaper/storage/CassandraStorage.java:153]: @adejanovski those reads and writes that are used multiple times within the one request+response codepath i've made explicitly `QUORUM`.

What is the penalty of QUORUM over LOCAL_QUORUM?

Reads and writes to cluster and schedules I feel should be kept "safe" as QUORUM.
But repair and segment i wonder if are safe to lower to LOCAL_QUORUM…?

**Pull request merged**